oioioi/mp/controllers.py

```diff
@@ -0,0 +1,198 @@
+
+    def registration_view(self, request):
+        participant = self._get_participant_for_form(request)
+
+        if 'mp_mpregistrationformdata' in request.session:
+            # pylint: disable=not-callable
+            form = self.form_class(request.session['mp_mpregistrationformdata'])
+            del request.session['mp_mpregistrationformdata']
+        else:
+            form = self.get_form(request, participant)
+        form.set_terms_accepted_text(self.get_terms_accepted_phrase())
+
+        if request.method == 'POST':
+            # pylint: disable=maybe-no-member
+            if form.is_valid():
+                participant, created = Participant.objects.get_or_create(
+                    contest=self.contest, user=request.user
+                )
+                self.handle_validated_form(request, form, participant)
+                if 'next' in request.GET:
+                    return safe_redirect(request, request.GET['next'])
+                else:
+                    return redirect('default_contest_view', contest_id=self.contest.id)
+        can_unregister = False
+        if participant:
+            can_unregister = self.can_unregister(request, participant)
+        context = {
+            'form': form,
+            'participant': participant,
+            'can_unregister': can_unregister,
+            'contest_name': self.contest.name
+        }
+        return TemplateResponse(request, self.registration_template, context)
+
+    def mixins_for_admin(self):
+        from oioioi.participants.admin import TermsAcceptedPhraseAdminMixin
+
+        return super(MPRegistrationController, self).mixins_for_admin() + (
+            TermsAcceptedPhraseAdminMixin,
+        )
+
+    def can_change_terms_accepted_phrase(self, request):
+        return not MPRegistration.objects.filter(
+            participant__contest=request.contest
+        ).exists()
+
+
+class MPContestController(ProgrammingContestController):
+    description = _("Mistrz Programowania")
+    create_forum = False
+
+    show_email_in_participants_data = True
+
+    def registration_controller(self):
+        return MPRegistrationController(self.contest)
+
+    def ranking_controller(self):
+        return MPRankingController(self.contest)
+
+    def update_user_result_for_problem(self, result):
+        """
+        """
+        submissions = Submission.objects.filter(
+            problem_instance=result.problem_instance,
+            user=result.user,
+            kind='NORMAL',
+            score__isnull=False,
+        )
+
+        if submissions:
+            best_submission = None
+            for submission in submissions:
+                ssm = SubmissionScoreMultiplier.objects.filter(
+                    contest=submission.problem_instance.contest,
+                )
+                
+                score = FloatScore(submission.score.value)
+                rtimes = self.get_round_times(None, submission.problem_instance.round)
```

Comment:
I understood this logic, but it is not trivial and should have some helpful comments added at least. Perhaps it could be turned into separate function to make this method more compact, like apply_submissionscoremultiplier